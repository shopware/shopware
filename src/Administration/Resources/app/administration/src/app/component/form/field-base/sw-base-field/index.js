import template from './sw-base-field.html.twig';
import './sw-base-field.scss';

const { Component } = Shopware;
const utils = Shopware.Utils;

Component.register('sw-base-field', {
    template,
    inheritAttrs: false,

    props: {
        name: {
            type: String,
            required: false,
            default: null
        },

        label: {
            type: String,
            required: false,
            default: null
        },

        helpText: {
            type: String,
            required: false,
            default: null
        },

        isInvalid: {
            type: Boolean,
            required: false,
            default: false
        },

        error: {
            type: [Object],
            required: false,
            default() {
                return null;
            }
        },

        disabled: {
            type: Boolean,
            required: false,
            default: false
        },

        required: {
            type: Boolean,
            required: false,
            default: false
        },

        isInherited: {
            type: Boolean,
            required: false,
            default: false
        },

        isInheritanceField: {
            type: Boolean,
            required: false,
            default: false
        },

        disableInheritanceToggle: {
            type: Boolean,
            required: false,
            default: false
        }
    },

    data() {
        return {
            id: utils.createId()
        };
    },

    computed: {
        identification() {
            if (this.name) {
                return this.name;
            }

            return `sw-field--${this.id}`;
        },

        hasLabel() {
            return !!this.label || !!this.helpText || this.isInheritanceField;
        },

        hasError() {
            return this.isInvalid || !!this.error;
        },

        swFieldClasses() {
            return {
                'has--error': this.hasError,
                'is--disabled': this.disabled,
                'is--inherited': this.isInherited
            };
        },

        swFieldLabelClasses() {
            return {
                'is--required': this.required
            };
        }
    }
});
