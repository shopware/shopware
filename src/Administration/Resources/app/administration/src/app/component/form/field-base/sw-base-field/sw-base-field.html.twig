{% block sw_base_field %}
    <div class="sw-field" :class="swFieldClasses">
        <div class="sw-field__label" v-if="hasLabel">
            <sw-inheritance-switch v-if="isInheritanceField"
                            :disabled="disableInheritanceToggle"
                            class="sw-field__inheritance-icon"
                            :isInherited="isInherited"
                            v-on="$listeners">
            </sw-inheritance-switch>

            <label v-if="label" :for="identification" :class="swFieldLabelClasses">
                {{ label }}
            </label>

            <sw-help-text class="sw-field__help-text" :text="helpText" v-if="helpText"></sw-help-text>
        </div>
        <slot name="sw-field-input" v-bind="{identification, error, disabled}"></slot>
        <sw-field-error :error="error"></sw-field-error>
    </div>
{% endblock %}
