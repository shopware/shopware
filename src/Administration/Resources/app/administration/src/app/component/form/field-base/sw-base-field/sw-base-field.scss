@import "~scss/variables";

.sw-field {
    width: 100%;
    margin-bottom: 22px;

    .sw-field__label {
        display: flex;
        line-height: 16px;
        font-size: 14px;
        margin-bottom: 8px;
        color: $color-kashmir;

        label {
            flex-grow: 1;
        }
    }

    .sw-field__label .is--required::after{
        content: "*";
        color: $color-shopware-blue
    }

    .sw-field__inheritance-icon {
        margin-right: 8px;
    }

    .sw-field__button-restore {
        color: $color-kashmir;
        padding: 0 8px;
        border: none;
        background: none;
        outline: none;
        -moz-appearance: none;
        -webkit-appearance: none;
        cursor: pointer;
    }

    .sw-field__help-text {
        align-self: center;
    }

    &.is--inherited {
        .sw-field__label {
            color: $color-purple;
        }
    }
}