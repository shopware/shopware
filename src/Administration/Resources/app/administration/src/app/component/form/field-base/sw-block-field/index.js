/**
 * @package admin
 */

import template from './sw-block-field.html.twig';
import './sw-block-field.scss';

const { Component } = Shopware;

/**
 * @deprecated tag:v6.6.0 - Will be private
 */
Component.register('sw-block-field', {
    template,
    inheritAttrs: false,

    props: {
        size: {
            type: String,
            required: false,
            default: 'default',
            validValues: ['small', 'medium', 'default'],
            validator(val) {
                return ['small', 'medium', 'default'].includes(val);
            },
        },
    },

    data() {
        return {
            hasFocus: false,
        };
    },

    computed: {
        swBlockSize() {
            return `sw-field--${this.size}`;
        },

        swBlockFieldClasses() {
            return [
                {
                    'has--focus': this.hasFocus,
                },
                this.swBlockSize,
            ];
        },
    },

    methods: {
        setFocusClass() {
            this.hasFocus = true;
        },

        removeFocusClass() {
            this.hasFocus = false;
        },
    },
});
