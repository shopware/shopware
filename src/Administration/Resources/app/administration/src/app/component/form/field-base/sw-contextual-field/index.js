/**
 * @package admin
 */

import template from './sw-contextual-field.html.twig';
import './sw-contextual-field.scss';

const { Component } = Shopware;

/**
 * @deprecated tag:v6.6.0 - Will be private
 */
Component.register('sw-contextual-field', {
    template,
    inheritAttrs: false,

    computed: {
        hasPrefix() {
            return this.$scopedSlots.hasOwnProperty('sw-contextual-field-prefix')
                && this.$scopedSlots['sw-contextual-field-prefix']({}) !== undefined;
        },

        hasSuffix() {
            return this.$scopedSlots.hasOwnProperty('sw-contextual-field-suffix')
                && this.$scopedSlots['sw-contextual-field-suffix']({}) !== undefined;
        },
    },
});
