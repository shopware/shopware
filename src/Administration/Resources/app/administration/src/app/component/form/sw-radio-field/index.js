import template from './sw-radio-field.html.twig';
import './sw-radio-field.scss';

const { Component, Mixin } = Shopware;

/**
 * @public
 * @description radio input field.
 * @status ready
 * @example-type static
 * @component-example
 * <sw-radio-field
 *      label="Radio field example"
 *      bordered
 *      :options="[
 *          {'value': 'value1', 'name': 'Label #1'},
 *          {'value': 'value2', 'name': 'Label #2'},
 *          {'value': 'value3', 'name': 'Label #3'},
 *          {'value': 'value4', 'name': 'Label #4'},
 *          {'value': 'value5', 'name': 'Label #5'}
 * ]"></sw-radio-field>
 */
Component.register('sw-radio-field', {
    template,
    inheritAttrs: false,

    mixins: [
        Mixin.getByName('sw-form-field'),
        Mixin.getByName('remove-api-error'),
    ],

    model: {
        prop: 'value',
        event: 'change',
    },

    props: {
        bordered: {
            type: Boolean,
            required: false,
            default: false,
        },

        block: {
            type: Boolean,
            required: false,
            default: false,
        },

        description: {
            type: String,
            required: false,
            default: null,
        },

        options: {
            type: Array,
            required: false,
            default: () => {
                return [];
            },
        },
        // FIXME: add type and default attribute to property
        // eslint-disable-next-line vue/require-prop-types, vue/require-default-prop
        value: {
            required: false,
        },
    },

    computed: {
        classes() {
            return [{
                'sw-field--radio-bordered': this.bordered,
                'sw-field--radio-block': this.block,
            }];
        },
        currentIndex() {
            const foundIndex = this.options.findIndex((item) => item.value === this.value);

            if (foundIndex < 0) {
                console.warn(`Given value "${this.value}" does not exists in given options`);
            }

            return foundIndex;
        },
    },

    methods: {
        onChange(event) {
            const selectedIndex = event.target.value;

            if (this.options[selectedIndex] === undefined) {
                console.warn(`Selected index "${this.value}" does not exists in given options`);
            }

            this.$emit('change', this.options[selectedIndex].value);
        },
    },
});
