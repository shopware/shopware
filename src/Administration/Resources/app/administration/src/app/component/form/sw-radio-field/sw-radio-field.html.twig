{% block sw_radio_field %}
<sw-base-field
    class="sw-field--radio"
    :class="classes"
    :name="formFieldName"
    v-bind="$attrs"
>
    <template #sw-field-input="{ identification, helpText, error, disabled }">
        <sw-help-text
            v-if="helpText"
            :text="helpText"
        />

        <div class="sw-field__radio-group">
            <div
                v-for="(option, index) in options"
                :key="index"
                class="sw-field__radio-option group-bordered"
                :class="{ 'sw-field__radio-option-checked': currentIndex === index }"
            >

                <div class="sw-field__radio-input">
                    <input
                        :id="`${identification}-${index}`"
                        type="radio"
                        :name="identification"
                        :value="index"
                        :disabled="disabled || option.disabled"
                        :checked="currentIndex === index"
                        @change="onChange"
                    >
                    <div class="sw-field__radio-state"></div>
                </div>

                <label :for="`${identification}-${index}`">
                    <span>{{ option.name }}</span>
                    {% block sw_radio_field_help_text %}
                    <sw-help-text
                        v-if="option.helpText"
                        :text="option.helpText"
                    />
                    {% endblock %}

                    <div
                        v-if="option.description"
                        class="sw-field__radio-option-description"
                    >
                        {{ option.description }}
                    </div>
                </label>

                <slot
                    :name="`custom-field-${option.value}`"
                    v-bind="{ option, disabled, checked: currentIndex === index }"
                ></slot>
            </div>
        </div>

        <div
            v-if="description"
            class="sw-field__radio-description"
        >
            {{ description }}
        </div>
    </template>

    <template #label>
        <slot name="label"></slot>
    </template>
</sw-base-field>
{% endblock %}
