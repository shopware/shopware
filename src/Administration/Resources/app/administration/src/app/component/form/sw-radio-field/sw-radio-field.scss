@import "~scss/variables";

$sw-field--radio-color-border: $color-gray-300;

.sw-field--radio {
    label ~ .sw-field__radio-group,
    .sw-help-text ~ .sw-field__radio-group {
        margin-top: 12px;
    }

    .sw-field__radio-option {
        display: grid;
        grid-template-columns: 16px 1fr 16px;
        grid-column-gap: 8px;
        margin-bottom: 12px;

        &.is--disabled {
            background-color: $color-gray-50;
        }

        label {
            font-weight: normal;
            font-size: $font-size-small;
            color: $color-darkgray-200;

            .sw-help-text {
                margin-left: 5px;
            }
        }

        &:hover,
        &:hover > * {
            cursor: pointer;
        }

        &.is--disabled:hover,
        &.is--disabled:hover > * {
            cursor: not-allowed;
        }
    }

    &.sw-field--radio-bordered .sw-field__radio-group {
        border-radius: 4px;
        border: 1px solid $sw-field--radio-color-border;
        padding: 12px 16px;
        margin-bottom: 0;
    }

    .sw-field__radio-input {
        width: 16px;
        height: 16px;
        position: relative;

        input[type="radio"] {
            opacity: 0;
            display: block;
            width: 100%;
            height: 100%;
            position: absolute;
            top: 0;
            left: 0;
            border: 0 none;
            background: none;
            -webkit-appearance: none;
            cursor: pointer;
            z-index: 2;

            &:disabled ~ .sw-field__radio-state {
                background: $color-gray-100;
                border-color: $color-gray-300;
                color: lighten($color-darkgray-200, 40%);
            }

            &:checked ~ .sw-field__radio-state {
                border: 5px solid $color-shopware-brand-500;
            }

            &:checked:disabled ~ .sw-field__radio-state {
                background: $color-gray-300;
                border: 1px solid $color-gray-300;
                box-shadow: inset 0 0 0 4px $color-gray-100;
                color: lighten($color-darkgray-200, 40%);
            }
        }

        .sw-field__radio-state {
            position: absolute;
            width: 100%;
            height: 100%;
            z-index: 1;
            background: $color-white;
            color: $color-darkgray-200;
            border: 1px solid $color-gray-300;
            border-radius: 8px;
        }
    }

    &.has--error {
        input[type="radio"]:checked ~ .sw-field__radio-state {
            border: 5px solid $color-crimson-500;
        }

        input[type="radio"]:checked ~ .sw-field__label {
            color: $color-crimson-500;
        }
    }

    &.is--disabled {
        .sw-field__radio-input {
            input[type="radio"] {
                cursor: not-allowed;
            }
        }
    }

    .sw-field__radio-option-label {
        &.error--selection {
            color: $color-crimson-500;
        }
    }

    .sw-field__radio-option-description {
        margin-top: 8px;
        color: $color-darkgray-200;
        font-size: $font-size-small;
        font-weight: normal;
        line-height: 22px;
    }

    .sw-field__radio-description {
        color: $color-gray-500;
        font-size: $font-size-small;
        line-height: 22px;
        padding-top: 8px;
    }

    &.sw-field--radio-block {
        .sw-field__label {
            font-size: $font-size-default;
            font-weight: bold;
            line-height: 22px;
            margin-bottom: 20px;
        }

        .sw-field__radio-option {
            transition: border-color ease-in 100ms;
            border: 1px solid $color-gray-200;
            border-radius: $border-radius-default;
            padding: 12px 19px;

            &:not(.is--disabled):hover {
                border-color: $color-shopware-brand-500;
            }
        }

        .sw-field__radio-option label {
            font-size: $font-size-default;
            font-weight: $font-weight-semi-bold;
            line-height: 22px;
            padding-left: 12px;
        }

        .sw-field__radio-input {
            margin-top: 5px;
        }

        .sw-field__radio-option.sw-field__radio-option-checked:not(.is--disabled) {
            background: rgba($color-shopware-brand-500, 0.05);
            border: 1px solid $color-shopware-brand-500;
        }
    }
}
