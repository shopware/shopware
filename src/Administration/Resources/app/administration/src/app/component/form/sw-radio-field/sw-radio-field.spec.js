/**
 * @package admin
 */

import { mount } from '@vue/test-utils';
import 'src/app/component/form/field-base/sw-base-field';
import 'src/app/component/form/sw-radio-field';

const createWrapper = async () => {
    const baseComponent = {
        template: `
            <sw-radio-field :options="options" v-model:value="currentValue" :block="block" :description="description">
                <template #custom-field-1="{ option, disabled, checked }">
                    <input id="custom-slot" type="text" :disabled="disabled || !checked">
                    <p>Custom slot</p>
                </template>
            </sw-radio-field>
        `,

        data() {
            return {
                options: [
                    { value: 1, name: 'option 1' },
                    { value: 2, name: 'option 2' },
                    { value: 3, name: 'option 3' },
                ],
                currentValue: 1,
                block: false,
                description: null,
            };
        },
    };

    const wrapper = mount(baseComponent, {
        global: {
            stubs: {
                'sw-radio-field': await wrapTestComponent('sw-radio-field', {
                    sync: true,
                }),
                'sw-base-field': await wrapTestComponent('sw-base-field'),
                'sw-field-error': {
                    template: '<div></div>',
                },
                'sw-help-text': true,
                'sw-inheritance-switch': true,
                'sw-ai-copilot-badge': true,
            },
        },
    });

    await flushPromises();

    return wrapper;
};

describe('components/form/sw-radio-field', () => {
    it('should renders correctly with children slot', async () => {
        const wrapper = await createWrapper();

        expect(wrapper.find('[id="custom-slot"]').exists()).toBe(true);
        expect(wrapper.find('p').wrapperElement).toHaveTextContent('Custom slot');
    });

    it('should render number of children equal to option props', async () => {
        const wrapper = await createWrapper();

        const radioInputs = wrapper.findAll('.sw-field__radio-option');
        expect(radioInputs).toHaveLength(3);
    });

    it('should pass slot scoped to children slot correctly', async () => {
        const wrapper = await createWrapper();
        const customSlot = wrapper.find('#custom-slot');

        await wrapper.setData({ currentValue: 1 });
        await wrapper.vm.$nextTick();

        expect(customSlot.wrapperElement).toBeEnabled();

        await wrapper.setData({ currentValue: 2 });
        await wrapper.vm.$nextTick();

        expect(customSlot.wrapperElement).toBeDisabled();
    });

    it('should render description block', async () => {
        const wrapper = await createWrapper();

        let description = wrapper.find('.sw-field__radio-description');
        expect(description.exists()).toBe(false);

        await wrapper.setData({ description: 'Lorem ipsum' });

        description = wrapper.find('.sw-field__radio-description');
        expect(description.exists()).toBe(true);
    });

    it('should render description of options', async () => {
        const wrapper = await createWrapper();

        let optionDescription = wrapper.find('.sw-field__radio-option-description');
        expect(optionDescription.exists()).toBe(false);

        await wrapper.setData({
            options: [
                { value: 1, name: 'option 1', description: 'option 1' },
                { value: 2, name: 'option 2', description: 'option 2' },
                { value: 3, name: 'option 3', description: 'option 3' },
            ],
        });

        optionDescription = wrapper.find('.sw-field__radio-option-description');
        expect(optionDescription.exists()).toBe(true);
    });

    it('should show the label from the property', async () => {
        const wrapper = mount(await wrapTestComponent('sw-radio-field', { sync: true }), {
            props: {
                label: 'Label from prop',
            },
            global: {
                stubs: {
                    'sw-base-field': await wrapTestComponent('sw-base-field'),
                    'sw-field-error': {
                        template: '<div></div>',
                    },
                    'sw-help-text': true,
                    'sw-inheritance-switch': true,
                    'sw-ai-copilot-badge': true,
                },
            },
        });

        await flushPromises();

        expect(wrapper.find('label').text()).toBe('Label from prop');
    });

    it('should show the value from the label slot', async () => {
        const wrapper = mount(await wrapTestComponent('sw-radio-field', { sync: true }), {
            props: {
                label: 'Label from prop',
            },
            global: {
                stubs: {
                    'sw-base-field': await wrapTestComponent('sw-base-field'),
                    'sw-field-error': {
                        template: '<div></div>',
                    },
                    'sw-help-text': true,
                    'sw-inheritance-switch': true,
                    'sw-ai-copilot-badge': true,
                },
            },
            slots: {
                label: '<template>Label from slot</template>',
            },
        });

        await flushPromises();

        expect(wrapper.find('label').text()).toBe('Label from slot');
    });
});
