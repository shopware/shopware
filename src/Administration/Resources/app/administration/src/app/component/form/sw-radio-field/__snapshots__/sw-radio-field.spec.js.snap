// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/sw-radio-field should renders correctly with children slot 1`] = `
<div
  class="sw-field sw-field--radio"
>
  <!---->
   
  <!---->
   
  <div
    class="sw-field__radio-group"
  >
    <div
      class="sw-field__radio-option group-bordered sw-field__radio-option-checked"
    >
      <div
        class="sw-field__radio-input"
      >
        <input
          id="sw-field--currentValue-0"
          name="sw-field--currentValue"
          type="radio"
          value="0"
        />
         
        <div
          class="sw-field__radio-state"
        />
      </div>
       
      <label
        class="sw-field__radio-option-label"
        for="sw-field--currentValue-0"
      >
        <span>
          option 1
        </span>
         
        <!---->
         
        <!---->
      </label>
       
      <input
        id="custom-slot"
        type="text"
      />
       
      <p>
        Custom slot
      </p>
    </div>
    <div
      class="sw-field__radio-option group-bordered"
    >
      <div
        class="sw-field__radio-input"
      >
        <input
          id="sw-field--currentValue-1"
          name="sw-field--currentValue"
          type="radio"
          value="1"
        />
         
        <div
          class="sw-field__radio-state"
        />
      </div>
       
      <label
        class="sw-field__radio-option-label"
        for="sw-field--currentValue-1"
      >
        <span>
          option 2
        </span>
         
        <!---->
         
        <!---->
      </label>
       
    </div>
    <div
      class="sw-field__radio-option group-bordered"
    >
      <div
        class="sw-field__radio-input"
      >
        <input
          id="sw-field--currentValue-2"
          name="sw-field--currentValue"
          type="radio"
          value="2"
        />
         
        <div
          class="sw-field__radio-state"
        />
      </div>
       
      <label
        class="sw-field__radio-option-label"
        for="sw-field--currentValue-2"
      >
        <span>
          option 3
        </span>
         
        <!---->
         
        <!---->
      </label>
       
    </div>
  </div>
   
  <!---->
   
  <div />
   
  <!---->
</div>
`;
